### PR TITLE
Swipe View Handler

### LINKED ISSUE #3932: SwipeView Handlers

### LINKED ISSUE #4154: SwipeView isn't working
### Description

    <CollectionView ItemsSource="{Binding Items}"
                    SelectionMode="Single"
                    SelectedItem="{Binding SelectedItem, Mode=TwoWay}"
                    SelectionChangedCommand="{Binding SelectCommand}"
                    Margin="0,12">
        <CollectionView.ItemTemplate>
            <DataTemplate>
                <SwipeView x:DataType="business:ContactGroupBusiness">
                    <SwipeView.RightItems>
                        <SwipeItem IconImageSource="{StaticResource Remove}"
                                       Invoked="OnRemove"/>
                    </SwipeView.RightItems>

                    
                    <Frame Margin="9,3">
                        <Label Text="{Binding Model.Name, Mode=OneWay}"
                                       Style="{StaticResource TitleStyle}"/>
                    </Frame>
                </SwipeView>
            </DataTemplate>
        </CollectionView.ItemTemplate>
    </CollectionView>

### Steps to Reproduce

1. Create a new Content Page;
2. Add Collection View with Swipe View

### Version with bug

Preview 11

### Last version that worked well

Unknown/Other

### Affected platforms

Android, I was *not* able test on other platforms

### Affected platform versions

Android 12

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###

SwipeViewHandler

### TODO Later PRs
- Log any bugs you find that aren't related to crashing
- AutomationId will get fixed up in a separate PR

### Additions made ###

```C#
public interface ISwipeView : IContentView
{
	public double Threshold { get; }

	public ISwipeItems LeftItems { get; }

	public ISwipeItems RightItems { get; }

	public ISwipeItems TopItems { get; }

	public ISwipeItems BottomItems { get; }

	public bool IsOpen { get; set; }

	public SwipeTransitionMode SwipeTransitionMode { get; }

	public void SwipeStarted(SwipeViewSwipeStarted swipeStarted);

	public void SwipeChanging(SwipeViewSwipeChanging swipeChanging);

	public void SwipeEnded(SwipeViewSwipeEnded swipeEnded);

	public void RequestOpen(SwipeViewOpenRequest swipeOpenRequest);

	public void RequestClose(SwipeViewCloseRequest swipeCloseRequest);
}
```

```C#
namespace Microsoft.Maui
{
	public interface IMenuElement : IElement, IImageSourcePart, IText
	{
		Paint? Background { get; }
		bool IsEnabled { get; }
		Visibility Visibility { get; }
	}

	public interface ISwipeItemMenuItem : IMenuElement, ISwipeItem
	{
	}

	public interface ISwipeItem : IElement
	{	
		void OnInvoked();

		string AutomationId { get; }
	}


	public interface ISwipeItems : IList<ISwipeItem>
	{
		public SwipeMode Mode { get; }

		public SwipeBehaviorOnInvoked SwipeBehaviorOnInvoked { get; }
	}

	public interface ISwipeItemView : IContentView, ISwipeItem
	{
	}
}
```

```C#
namespace Microsoft.Maui.Handlers
{
	public partial interface ISwipeViewHandler : IViewHandler
	{
		ISwipeView TypedVirtualView { get; }
		NativeView TypedNativeView { get; }
	}
}
```

- class SwipeViewHandler
- class SwipeItemMenuItemHandler
- class SwipeItemViewHandler

### PR Checklist ###

- [ ] Targets the correct branch 
- [ ] Tests are passing (or failures are unrelated)
- [ ] Targets a single property for a single control (or intertwined few properties)
- [ ] Adds the property to the appropriate interface
- [ ] Avoids any changes not essential to the handler property
- [ ] Adds the mapping to the PropertyMapper in the handler
- [ ] Adds the mapping method to the WinUI, Android, iOS, and Standard aspects of the handler
- [ ] Implements the actual property updates (usually in extension methods in the Platform section of Core)
- [ ] Tags ported renderer methods with [PortHandler]
- [ ] Adds an example of the property to the sample project (MainPage)
- [ ] Adds the property to the stub class
- [ ] Implements basic property tests in DeviceTests

#### Does this PR touch anything that might affect accessibility?
- [ ] Does this PR introduce a new control? (If yes, add an example using SemanticProperties to the SemanticsPage)
- [ ] APIs that modify focusability?
- [ ] APIs that modify any text property on a control?
- [ ] Does this PR modify view nesting or view arrangement in anyway?
- [ ] Is there the smallest possibility that your PR will change accessibility? 
- [ ] I'm not sure, please help me

If any of the above checkboxes apply to your PR, then the PR will need to provide testing to demonstrate that accessibility still works.

### CODE REVIEW

**Review by jsuarezruiz — COMMENTED** [review]
Could you rebase to fix the conflicts?. Thanks!.

File: src/Controls/samples/Controls.Sample/Maui.Controls.Sample-net6.csproj [hunk 1]
```diff
@@ -62,6 +62,10 @@
     </MauiXaml>
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="Pages\Controls\SwipeViewGalleries\" />
+  </ItemGroup>
+
   <Import Condition=" '$(UseMaui)' != 'true' " Project="..\..\..\BlazorWebView\src\Maui\build\Microsoft.AspNetCore.Components.WebView.Maui.targets" />
```
Comment (jsuarezruiz) [inline src/Controls/samples/Controls.Sample/Maui.Controls.Sample-net6.csproj:69]: This should not be necessary.

File: src/Core/src/Handlers/SwipeView/SwipeItemMenuItemHandler.Android.cs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+			handler.NativeView.Visibility = view.Visibility.ToNativeVisibility();
+		}
+
+		static Color? GetSwipeItemColor(Color? backgroundColor)
+		{
+			if (backgroundColor == null)
+				return null;
+
+			var luminosity = 0.2126f * backgroundColor.Red + 0.7152f * backgroundColor.Green + 0.0722f * backgroundColor.Blue;
```
Comment (jsuarezruiz) [inline src/Core/src/Handlers/SwipeView/SwipeItemMenuItemHandler.Android.cs:131]: I know this code comes from Xamarin.Forms but could be a good moment to create an extension method with this.

File: src/Core/src/Core/ISwipeItemView.cs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Microsoft.Maui
+{
+	public interface ISwipeItemView : IContentView, ISwipeItem
```
Comment (jsuarezruiz) [inline src/Core/src/Core/ISwipeItemView.cs:3]: I think we should have a team meeting reviewing this point. SwipeItemView is  SwipeItem that could contain a View (anything). Is implemented in Android and iOS in Xamarin.Forms but NOT in Windows. In Windows we use https://docs.microsoft.com/es-es/windows/winui/api/microsoft.ui.xaml.controls.swipecontrol?view=winui-3.0 and not allow this scenario. So we have two approaches:
- Change the Windows implementation creating our own MauiSwipeView and allow to use SwipeItemView everywhere.
- Remove the SwipeItemView support and align the functionality to common denominator.

Comment (PureWeen) [reply src/Core/src/Core/ISwipeItemView.cs:3]: Can we do option one but create a tracking issue for it?
I don't think we want to diminish the behavior here for users upgrading from XF to .NET MAUI

File: src/Core/src/Handlers/SwipeView/SwipeViewHandler.Android.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+		public static void MapIsEnabled(ISwipeViewHandler handler, ISwipeView swipeView)
+		{
+			handler.TypedNativeView.UpdateIsSwipeEnabled(swipeView.IsEnabled);
+			ViewHandler.MapIsEnabled(handler, swipeView);
+		}
+
+		public static void MapBackground(ISwipeViewHandler handler, ISwipeView swipeView)
+		{
+			// TODO JAVIER is this right? I wasn't sure exactly how to combine the two
```
Comment (PureWeen) [inline src/Core/src/Handlers/SwipeView/SwipeViewHandler.Android.cs:46]: @jsuarezruiz  ?

File: src/Core/src/Core/ISwipeItemView.cs (hunk 3 above)
Comment (jsuarezruiz) [reply src/Core/src/Core/ISwipeItemView.cs:3]: I am going to create the issue to track it.

File: src/Core/src/Handlers/SwipeView/SwipeViewHandler.Android.cs (hunk 4 above)
Comment (jsuarezruiz) [reply src/Core/src/Handlers/SwipeView/SwipeViewHandler.Android.cs:46]: The logic in Xamarin.Forms was, apply the brush from Background and if is null, check and apply the color from the BackgroundColor property.

**Review by jsuarezruiz — COMMENTED** [review]
Have been reviewing the code and testing all the samples. There are some details in some cases I think not related with this code (For example: wrong size in Frame Content). Should we push to fix this stuff before merge or should we create issues?. Either way, I have created issues. We could associate them with this PR or upcoming ones.

Related issues:
https://github.com/dotnet/maui/issues/4118
https://github.com/dotnet/maui/issues/4119
https://github.com/dotnet/maui/issues/4120
https://github.com/dotnet/maui/issues/4121
https://github.com/dotnet/maui/issues/4122

There are other related issues like the CarouselView freeze and crash the App but is for an unrelated issue. I would love to review everything before merging. However, if we want to have the Handler in the iteration 13 we can approve this PR and continue iterating, checking the related bugs created.

**Review by jsuarezruiz — APPROVED** [review]

**Pull request merged**